pubsub/pubsub.go

```diff
@@ -0,0 +1,208 @@
+// Package pubsub implements a pub-sub model where there is a single publisher
+// (Server) and multiple subscribers.
+//
+// Though you can have multiple publishers by sharing a pointer to a server or
+// by giving the same channel to each publisher and publishing messages from
+// that channel.
+//
+// Clients subscribe for messages, which could be of any type, using a query. A
+// query can be anything as long as it conforms to the interface. See query
+// package.
+//
+// Server will be non-blocking for publishing messages if you provide an
+// OverflowStrategyDrop option. Note sending messages to subscribers is always
+// non-blocking.
+package pubsub
+
+import (
+	"errors"
+
+	cmn "github.com/tendermint/tmlibs/common"
+)
+
+type operation int
+
+const (
+	sub operation = iota
+	pub
+	unsub
+	shutdown
+)
+
+type overflowStrategy int
+
+const (
+	drop overflowStrategy = iota
+	wait
+)
+
+var (
+	ErrorOverflow = errors.New("Server overflowed")
+)
+
+type cmd struct {
+	op    operation
+	query Query
+	ch    chan<- interface{}
+	msg   interface{}
+	tags  map[string]interface{}
+}
+
+// Query defines an interface for a query to be used for subscribing.
+type Query interface {
+	Matches(tags map[string]interface{}) bool
+}
+
+// Server allows clients to subscribe/unsubscribe for messages, pubsling
+// messages with or without tags, and manages internal state.
+type Server struct {
+	cmn.BaseService
+
+	cmds             chan cmd
+	overflowStrategy overflowStrategy
+}
+
+// NewServer returns a new server. You can provide additional options such as
+// BufferCapacity(N) or OverflowStrategy{X}. See their documentation for
+// details.
+func NewServer(options ...func(*Server)) *Server {
+	s := &Server{overflowStrategy: wait}
+	s.BaseService = *cmn.NewBaseService(nil, "PubSub", s)
+
+	for _, option := range options {
+		option(s)
+	}
+
+	if s.cmds == nil { // if BufferCapacity was not set, create unbuffered channel
+		s.cmds = make(chan cmd)
+	}
+
+	return s
+}
+
+// BufferCapacity allows you to specify capacity for the internal servers
+// queue. Since the server given Y subscribers could only process X messages,
+// this option could be used to survive spikes (e.g. high amount of
+// transactions during peak hours).
+func BufferCapacity(cap int) func(*Server) {
+	return func(s *Server) {
+		if cap > 0 {
+			s.cmds = make(chan cmd, cap)
+		}
+	}
+}
+
+// OverflowStrategyDrop will tell the server to drop messages when it can't
+// process more messages.
+func OverflowStrategyDrop() func(*Server) {
+	return func(s *Server) {
+		s.overflowStrategy = drop
+	}
+}
+
+// OverflowStrategyWait will tell the server to block and wait for some time
+// for server to process other messages. Default strategy.
+func OverflowStrategyWait() func(*Server) {
+	return func(s *Server) {
+		s.overflowStrategy = wait
+	}
+}
+
+// Subscribe returns a channel on which messages matching the given query can
```

Comment:
No. It used to before. It is a question of control: who should be in control of a channel? and who should decide on the capacity of that channel? subscriber seems to be the more natural choice since only it knows how fast it will process events. 

like subscriber A wants `out` channel to be buffered (5) so it could read 5 events in a row and do something with them. 

Does it make sense?

ps. noticed this pattern here https://blog.golang.org/pipelines